sqlx-core/src/query_builder.rs

```diff
@@ -0,0 +1,141 @@
+use std::fmt::Display;
+
+use crate::arguments::Arguments;
+use crate::database::{Database, HasArguments};
+use crate::encode::Encode;
+use crate::query::Query;
+use crate::types::Type;
+use either::Either;
+use std::marker::PhantomData;
+
+pub struct QueryBuilder<'a, DB>
+where
+    DB: Database,
+{
+    query: String,
+    arguments: Option<<DB as HasArguments<'a>>::Arguments>,
+    variable_count: u16,
+}
+
+impl<'a, DB: Database> QueryBuilder<'a, DB>
+where
+    DB: Database,
+{
+    pub fn new(init: impl Into<String>) -> Self
+    where
+        <DB as HasArguments<'a>>::Arguments: Default,
+    {
+        QueryBuilder {
+            query: init.into(),
+            arguments: Some(Default::default()),
+            variable_count: 0,
+        }
+    }
+
+    pub fn push(&mut self, sql: impl Display) -> &mut Self {
+        self.query.push_str(&format!(" {}", sql));
```

Comment:
```suggestion
        // An error here is also highly unlikely but ignoring it is a code smell.
        write!(self.query, " {}", sql).expect("error formatting `sql`");
```

It is a bit weird to insert a space automatically, though. What if someone is trying to concatenate an identifier using this method, or is building a string literal where formatting matters? This violates the principle of least surprise.